### ISSUE TITLE
[Question] PooledUdpTransport with connected sockets

### ISSUE THREAD

**Issue by dv00d00** [justeat/JustEat.StatsD#85]
Addition of a pulled udp was very important and 20 times the throughput increase is fantastic.

I am wondering though if we could push it a bit further.

What if we will pool connected sockets, like this:

```csharp
        private readonly ConcurrentDictionary<IPEndPoint, Lazy<SimpleObjectPool<Socket>>> _pools =
            new ConcurrentDictionary<IPEndPoint, Lazy<SimpleObjectPool<Socket>>>();

...

        private SimpleObjectPool<Socket> GetPool(IPEndPoint endpoint)
        {
            var pool = _pools.GetOrAdd(
                endpoint,
                ep =>
                {
                    return new Lazy<SimpleObjectPool<Socket>>(
                        () =>
                        {
                            var result = new SimpleObjectPool<Socket>(
                                magic_number,
                                _ =>
                                {
                                    var socket = UdpTransport.CreateSocket();
                                    socket.Connect(ep);
                                    return socket;
                                });

                            return result;
                        }, LazyThreadSafetyMode.ExecutionAndPublication);
                });

            return pool.Value;
        }

// consumption

            var endpoint = _endpointSource.GetEndpoint();
            var pool = GetPool(endpoint);
            var socket = pool.Pop();
            socket.Send(data);

```

I've tried this on my machine:

 |Method |     Toolchain |       Mean |      Error |     StdDev |     Median | Scaled | Allocated |
|----------------- |-------------- |-----------:|-----------:|-----------:|-----------:|-------:|----------:|
| Send | .NET Core 2.1 | 474.578 us | 28.0948 us | 41.1810 us | 464.496 us |   1.00 |     400 B |
| SendWithPool | .NET Core 2.1 |  16.735 us |  0.2070 us |  0.2969 us |  16.696 us |   0.04 |     168 B |
| SendWithMegaPool | .NET Core 2.1 |   6.386 us |  0.1625 us |  0.2330 us |   6.308 us |   0.01 |       0 B |

**Comment by martincostello** [comment]
Interesting - though I'm slightly confused as to what's actually making this faster?

We generally only have one endpoint because our Graphite service is only in one place, so based on he benchmark that's currently in `master` I'm not quite following where the performance gain would have come from with more code being added.

**Comment by dv00d00** [comment]
`SendTo` copies the endpoint and does something else wich allocates 168 bytes + internally it is 2 native dll function calls (at least on Windows), basically, it does connect on every send.

While plain `Send` just transmits the data to established OS buffer

**Comment by martincostello** [comment]
Right, now I see - I was putting all of my focus onto the dictionary being added, so didn't see that `SendTo()` had changed to `Send()`.

This suggestion is actually quite similar to [this code](https://github.com/justeat/JustEat.StatsD/pull/80#issuecomment-405845206) from which #81 was based. That was based on the `2.1.0` that used `UdpClient`, so it was piggy-backing the fact that it maintained connection state, and was doing a similar thing regarding keeping the connections open and updating if the endpoint changed.

**Comment by AnthonySteele** [comment]
Nice! I think that we should do this, after the current PR. 

It can be simpler: endpoint changes are infrequent, and we are only interested in the current one. We don't expect to start using the previous one again so sockets for the previous endpoint(s) can be discarded, rather than kept in a dictionary.

**Comment by dv00d00** [comment]
+ Important side effect if we will choose to go `Send` route. `Send` already has a Span overload, while `SendTo` does not. 😅 Would justify span introduction quiet a lot

**Comment by martincostello** [comment]
We're looking at `PooledUdpTransport` being a non-breaking opt-in additional feature for `3.1.0` for now.

Any breaking-changes to the public surface would only be considered for a future `4.0.0` release.

**Comment by martincostello** [comment]
Resolved by #86.
